File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/AverageOfSubarraysOfSizeK.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class AverageOfSubarraysOfSizeK {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = averageOfSubarraysOfSizeK(a, k);
        for (var r : res) {
          System.out.printf("%.1f ", r);
        }
        System.out.println();
      }
    }
  }

  // O(n) / O(1)
  public static double[] averageOfSubarraysOfSizeK(int[] a, int k) {
    int n = a.length;
    double[] res = new double[n - k + 1];
    double sum = 0.0;
    int windowStart = 0;

    for (int windowEnd = 0; windowEnd < n; windowEnd++) {
      sum += a[windowEnd];

      if (windowEnd >= k - 1) {
        res[windowStart] = sum / (double)k;
        sum -= a[windowStart];
        windowStart++;
      }
    }

    return res;
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_3/average_of_subarrays_of_size_k.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace std;

// O(n) / O(1)
vector<double> averages(const vector<int> &a, int k) {
  int n = a.size();
  vector<double> res(n - k + 1);
  int window_start = 0, sum = 0;

  for (int window_end = 0; window_end < n; window_end++) {
    sum += a[window_end];

    if (window_end >= k - 1) {
      res[window_start] = (double)sum / (double)k;
      sum -= a[window_start];
      window_start++;
    }
  }

  return res;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = averages(a, k);
    for (double r : res) {
      cout << fixed << setprecision(1) << r << " ";
    }
    cout << "\n";
  }

  return 0;
}